packages/url/src/index.js

```diff
@@ -207,16 +208,25 @@ export function isValidQueryString( queryString ) {
  * @example
  * ```js
  * const fragment1 = getFragment( 'http://localhost:8080/this/is/a/test?query=true#fragment' ); // '#fragment'
- * const fragment2 = getFragment( 'https://wordpress.org#another-fragment?query=true' ); // '#another-fragment'
+ * const fragment2 = getFragment( 'https://wordpress.org#another-fragment?query=true' ); // '#another-fragment?query=true'
+ * const fragment3 = getFragment( 'https://wordpress.org' ); // undefined
  * ```
  *
  * @return {?string} The fragment part of the URL.
  */
 export function getFragment( url ) {
-	const matches = /^\S+?(#[^\s\?]*)/.exec( url );
-	if ( matches ) {
-		return matches[ 1 ];
+	const fragment = url
+		.replace( /^[^#]*/, '' )
+		.replace( /\s*$/, '' );
+
+	if ( '' === fragment ) {
+		return;
 	}
+
+	return encodeURI( fragment )
+		.replace( /%5B/gi, '[' )
```

Comment:
Can you explain for what reason these are here?

Reply:
Sure! In Firefox 67, this:

```js
new URL( 'https://example.com/some-page/?q=1#a hash-%with[spaces_and-öther characterß]' ).hash
```

returns this:

```js
"#a%20hash-%with[spaces_and-%C3%B6ther%20character%C3%9F]"
```

That is, it encoded almost all characters as `encodeURI` does, the only exceptions being `%`, `[`, and `]`. That's why I URI encoded the fragment and then reverted the scapes:

```js
encodeURI( "a hash-%with[spaces_and-öther characterß]" )
```

returns:

```js
"a%20hash-%25with%5Bspaces_and-%C3%B6ther%20character%C3%9F%5D"
```

which, after applying the replaces, becomes:

```js
"a%20hash-%with[spaces_and-%C3%B6ther%20character%C3%9F]"
```

That's why I applied those... Does this makes sense?

Reply:
Sort of. But rather than rely on some anecdotal observations of a specific browser, is there a specification from which we can draw on for the precise behavior to which we should conform?

In any case, it would be important to include sufficient inline code comments to make this behavior as clear as possible to the next maintainer.